cmd/autoscaler/main.go

```diff
@@ -201,8 +164,81 @@ func main() {
+	if err != nil {
+		log.Fatalf("Error loading logging configuration: %v", err)
+	}
+	loggingConfig, err := logging.NewConfigFromMap(loggingConfigMap)
+	if err != nil {
+		log.Fatalf("Error parsing logging configuration: %v", err)
+	}
+	return logging.NewLoggerFromConfig(loggingConfig, component)
+}
+
+func setupChannels() (<-chan struct{}, chan *autoscaler.StatMessage) {
+	stopCh := signals.SetupSignalHandler()
+	statsCh := make(chan *autoscaler.StatMessage, statsBufferLen)
+	return stopCh, statsCh
+}
+
+func setupClients(logger *zap.SugaredLogger, stopCh <-chan struct{}) (kubernetes.Interface, clientset.Interface, scale.ScalesGetter) {
+	cfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
+	if err != nil {
+		logger.Fatalw("Error building kubeconfig", zap.Error(err))
+	}
+
+	kubeClientSet, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalw("Error building kubernetes clientset", zap.Error(err))
+	}
+
+	if err := version.CheckMinimumVersion(kubeClientSet.Discovery()); err != nil {
+		logger.Fatalf("Version check failed: %v", err)
+	}
+
+	// This is based on how Kubernetes sets up its scale client based on discovery:
+	// https://github.com/kubernetes/kubernetes/blob/94c2c6c84/cmd/kube-controller-manager/app/autoscaling.go#L75-L81
+	restMapper := buildRESTMapper(kubeClientSet, stopCh)
+	scaleClient, err := scale.NewForConfig(cfg, restMapper, dynamic.LegacyAPIPathResolverFunc,
+		scale.NewDiscoveryScaleKindResolver(kubeClientSet.Discovery()))
+	if err != nil {
+		logger.Fatalw("Error building scale clientset", zap.Error(err))
+	}
+
+	servingClientSet, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalw("Error building serving clientset", zap.Error(err))
+	}
+
+	return kubeClientSet, servingClientSet, scaleClient
+}
+
+func setupAutoscalerConfig(logger *zap.SugaredLogger) *autoscaler.DynamicConfig {
+	rawConfig, err := configmap.Load("/etc/config-autoscaler")
+	if err != nil {
+		logger.Fatalw("Error reading autoscaler configuration", zap.Error(err))
+	}
+	dynConfig, err := autoscaler.NewDynamicConfigFromMap(rawConfig, logger)
+	if err != nil {
+		logger.Fatalw("Error parsing autoscaler configuration", zap.Error(err))
+	}
+	return dynConfig
+}
+
+func setupConfigWatcher(kc kubernetes.Interface, logger *zap.SugaredLogger, level zap.AtomicLevel, dynConfig *autoscaler.DynamicConfig) *configmap.InformedWatcher {
+	configMapWatcher := configmap.NewInformedWatcher(kc, system.Namespace())
+
+	// Watch the logging config map and dynamically update logging levels.
+	configMapWatcher.Watch(logging.ConfigName, logging.UpdateLevelFromConfigMap(logger, level, component))
+	// Watch the observability config map and dynamically update metrics exporter.
+	configMapWatcher.Watch(metrics.ObservabilityConfigName, metrics.UpdateExporterFromConfigMap(component, logger))
+	// Watch the autoscaler config map and dynamically update autoscaler config.
+	configMapWatcher.Watch(autoscaler.ConfigName, dynConfig.Update)
```

Comment:
Should the watch calls be inlined and closer to where they are wanted? E.g. the autoscaler watch close to the autoscalerConfig setup and the loglevel close to logger setup?

Reply:
Yes, I think it'd be ideal if we can have the watcher set up beforehand, and then watch the logger right after it is set up and watch the autoscaler config right after it is set up.  But there is a chain of dependency preventing us from doing that.  ConfigMapWatcher needs kubeClientSet, which needs logger.